lib/sdl_x11.cpp

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) 2010-2021, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-443271.
+//
+// This file is part of the GLVis visualization tool and library. For more
+// information and source code availability see https://glvis.org.
+//
+// GLVis is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#include "sdl_x11.hpp"
+
+#ifdef SDL_VIDEO_DRIVER_X11
+
+#include <vector>
+#include <array>
+#include <iostream>
+
+#ifndef __EMSCRIPTEN__
+#include <SDL2/SDL_syswm.h>
+#else
+#include <SDL_syswm.h>
+#endif
+
+#ifdef SDL_VIDEO_DRIVER_X11_XINPUT2
+#include <X11/extensions/XInput2.h>
+#endif // SDL_VIDEO_DRIVER_X11_XINPUT2
+
+using namespace std;
+
+void SdlX11Platform::RegisterWindow(SDL_Window* window)
+{
+   SDL_SysWMinfo sysinfo;
+   SDL_VERSION(&sysinfo.version);
+   if (!SDL_GetWindowWMInfo(window, &sysinfo))
+   {
+      cerr << "Error: unable to get window manager information for the "
+           << "current window." << endl;
+      return;
+   }
+   if (sysinfo.subsystem != SDL_SYSWM_X11)
+   {
+      cerr << "Error: created SDL window is not an X11 window." << endl;
+      return;
+   }
+   Display *disp = sysinfo.info.x11.display;
+   Window wnd = sysinfo.info.x11.window;
+
+   display_fds.emplace(window, ConnectionNumber(disp));
+
+#ifdef SDL_VIDEO_DRIVER_X11_XINPUT2
+   // Disable XInput extension events since they are generated even outside
+   // the GLVis window.
+   Window root_win = DefaultRootWindow(disp);
+   unsigned char mask[4] = {0,0,0,0};
+   XIEventMask event_mask;
+   event_mask.deviceid = XIAllMasterDevices;
+   event_mask.mask_len = sizeof(mask);
+   event_mask.mask = mask;
+#ifdef SDL_VIDEO_DRIVER_X11_DYNAMIC_XINPUT2
+   const char Xi_lib[] = SDL_VIDEO_DRIVER_X11_DYNAMIC_XINPUT2;
+#else
+   const char Xi_lib[] = "libXi.so";
+#endif
+   typedef int (*XISelectEvents_ptr)(Display *, Window, XIEventMask *, int);
+   XISelectEvents_ptr XISelectEvents_ = NULL;
+   void *lib = SDL_LoadObject(Xi_lib);
+   if (lib != NULL)
+   {
+      XISelectEvents_ =
+         (XISelectEvents_ptr)SDL_LoadFunction(lib, "XISelectEvents");
+   }
+   if (XISelectEvents_ == NULL)
+   {
+      cerr << "Error accessing XISelectEvents!" << endl;
+      exit(EXIT_FAILURE);
+   }
+   if (XISelectEvents_(disp, root_win, &event_mask, 1) != Success)
+   {
+      cerr << "Failed to disable XInput on the default root window!" << endl;
+   }
+   if (XISelectEvents_(disp, wnd, &event_mask, 1) != Success)
+   {
+      cerr << "Failed to disable XInput on the current window!" << endl;
```

Comment:
should these error blocks also `exit(EXIT_FAILURE)` ?

Reply:
Yeah, I think if we can't disable XInput extension events, it just means our event loop gets woken up more.